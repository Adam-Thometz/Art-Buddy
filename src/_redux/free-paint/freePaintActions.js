import { createAction } from "@reduxjs/toolkit";

export const setColor = createAction('setColor');
export const setStencil = createAction('setStencil');
export const setIsErasing = createAction('setIsErasing');
export const clearGame = createAction('clearGame');