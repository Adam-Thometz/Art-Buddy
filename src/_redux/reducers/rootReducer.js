// import { combineReducers } from 'redux';
import { configureStore } from '@reduxjs/toolkit';

import mainSettingsReducer from './mainSettingsReducer';
import wordToMusicDecoderReducer from './wordToMusicDecoderReducer';
import sequenceMakerReducer from './sequenceMakerReducer';
import scoreKeeperReducer from './scoreKeeperReducer';
import instrumentIdReducer from './instrumentIdReducer';

const rootReducer = configureStore({
  reducer: {
    mainSettings: mainSettingsReducer,
    wordToMusic: wordToMusicDecoderReducer,
    sequenceMaker: sequenceMakerReducer,
    scoreKeeper: scoreKeeperReducer,
    instrumentId: instrumentIdReducer
  }
});

export default rootReducer;