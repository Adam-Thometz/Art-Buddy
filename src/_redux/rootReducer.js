import { configureStore } from '@reduxjs/toolkit';

import mainSettingsReducer from './settings/mainSettingsReducer';
import musicDecoderReducer from './music-decoder/musicDecoderReducer';
import jumpIntoRhythmReducer from './jump-into-rhythm/jumpIntoRhythmReducer';
import sequenceMakerReducer from './sequence-maker/sequenceMakerReducer';
import idListeningSkillsTestReducer from './instrument-id/listening-skills/idListeningSkillsReducer';
import idSongMakerReducer from './instrument-id/song-maker/idSongMakerReducer';
import freePaintReducer from './free-paint/freePaintReducer';
import colorTheoryReducer from './color-theory/colorTheoryReducer';
import scoreKeeperReducer from './score-keeper/scoreKeeperReducer';

const rootReducer = configureStore({
  reducer: {
    mainSettings: mainSettingsReducer,
    musicDecoder: musicDecoderReducer,
    sequenceMaker: sequenceMakerReducer,
    listeningSkillsTest: idListeningSkillsTestReducer,
    songMaker: idSongMakerReducer,
    scoreKeeper: scoreKeeperReducer,
    jumpIntoRhythm: jumpIntoRhythmReducer,
    freePaint: freePaintReducer,
    colorTheory: colorTheoryReducer
  }
});

export default rootReducer;