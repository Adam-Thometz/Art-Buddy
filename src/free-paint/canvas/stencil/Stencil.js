import React from 'react';

import { useSelector } from 'react-redux';

import './Stencil.css';

const Stencil = () => {
  const { stencil } = useSelector(state => state.freePaint);

  // const handleClearDisplay = () => dispatch(setDisplay(null));

  return (
    <div className='Stencil'>
      {stencil.startsWith('data:image')
        ? <img className='Stencil-shape' src={stencil} alt='' />
        : <span className='Stencil-text'>{stencil}</span>
      }
    </div>
  );
};

export default Stencil;