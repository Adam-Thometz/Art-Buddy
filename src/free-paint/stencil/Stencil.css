.Stencil {
  position: absolute;
  background-color: #F8F8F8;
  top: 12%;
  right: 6%;
  border: 1px solid #000000;
  border-radius: 10%;
  width: 40%;
  height: 70vh;
  display: flex;
  justify-content: center;
  align-items: center;
}

.Stencil-text {
  font-family: Arial;
  font-size: 35rem;
  color: white;
  -webkit-text-stroke-width: 5px;
  -webkit-text-stroke-color: #000000;
  /* To be used for scaling */
  /* -webkit-transform: scale(2,1); Safari and Chrome */
  /* -moz-transform: scale(2,1); Firefox */
  /* -ms-transform: scale(2,1); IE 9 */
  /* -o-transform: scale(2,1); Opera */
  /* transform: scale(2,1); W3C */
}

.Stencil-clear {
  height: 40px;
  width: 40px;
  position: absolute;
  top: 5%;
  right: 5%;
  cursor: pointer;
}

.Stencil-shape {
  height: 80%;
  width: 80%;
}