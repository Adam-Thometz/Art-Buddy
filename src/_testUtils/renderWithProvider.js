import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

// import renderWithRouter from './renderWithRouter';
import rootReducer from '../_redux/reducers/rootReducer';

function withProvider({ children, store }) {
  return (
    <Provider store={store}>
      <MemoryRouter>
        {children}
      </MemoryRouter>
    </Provider>
  )
};

function renderWithProvider(component, options) {
  const store = (options && options.store)
    ? options.store
    : rootReducer;
  return render(component, { wrapper: (args) =>
    withProvider({
      ...args,
      store
    }), ...options
  });
};

export default renderWithProvider;