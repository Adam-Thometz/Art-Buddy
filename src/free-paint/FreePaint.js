import React, { useEffect } from 'react';
import useVisited from '_hooks/useVisited';

import { useSelector, useDispatch } from 'react-redux';
import { clearGame } from '_redux/free-paint/freePaintActions';
import { changeCurrGame } from '_redux/_general/generalActions';

import WindowNavbar from '_components/window-nav/WindowNavbar';
import Instructions from '_components/instructions/Instructions';
import ControlBar from './control-bar/ControlBar';
import Canvas from './canvas/Canvas';
import Stencil from './stencil/Stencil';

import { freePaint } from '_data/_activities/activityList';

const FreePaint = () => {
  const [hasVisited] = useVisited(freePaint.lsKey);
  const { currGame } = useSelector(state => state.general);
  const { stencil } = useSelector(state => state.freePaint);
  const dispatch = useDispatch();
  
  useEffect(() => {
    dispatch(changeCurrGame(freePaint));
    return () => {
      dispatch(clearGame())
      dispatch(changeCurrGame({}));
    };
  }, [dispatch]);

  return (<>
    <WindowNavbar page={currGame.name} />
    {!hasVisited ? <Instructions game={currGame} /> : <>
      <ControlBar />
      <Canvas />
      {stencil ? <Stencil /> : null}
    </>}
  </>);
};

export default FreePaint;