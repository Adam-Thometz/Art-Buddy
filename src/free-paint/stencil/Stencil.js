import React, { useState, useRef, useEffect } from 'react';
import Draggable from 'react-draggable';

import { useSelector, useDispatch } from 'react-redux';
import { setStencil } from '_redux/free-paint/freePaintActions';

import './Stencil.css';

import close from '_media/general-icons/close.png';
import resizeIcon from '_media/free-paint/resize.png';
import getNewSize from '_utils/free-paint/getNewSize';

const Stencil = () => {
  const { stencil, isEditingStencil } = useSelector(state => state.freePaint);
  const dispatch = useDispatch();
  const stencilRef = useRef();

  const [resizeInfo, setResizeInfo] = useState({ active: false, x: '', y: '' });
  const [size, setSize] = useState({ width: '', height: '', defaultWidth: '', defaultHeight: '' });

  useEffect(() => {
    const width = stencilRef.current.clientWidth;
    const height = stencilRef.current.clientHeight;
    setSize({ width, height, defaultWidth: width, defaultHeight: height });
  }, [stencilRef]);
  
  const beginResize = e => setResizeInfo({ active: true, x: e.clientX, y: e.clientY });
  
  const handleResize = e => {
    if (!resizeInfo.active) return;
    const { x, y } = resizeInfo;
    const { clientX, clientY } = e;
    const { height, width } = getNewSize({ x, y, clientX, clientY, currSize: size })
    setResizeInfo({ ...resizeInfo, x: clientX, y: clientY });
    setSize({ ...size, height, width });
  };
  
  const endResize = () => setResizeInfo({ ...resizeInfo, active: false });
  
  const handleClearDisplay = () => dispatch(setStencil(null));

  const style = {
    width: `${size.width}px`,
    height: `${size.height}px`,
    zIndex: isEditingStencil ? 5 : 3,
    '--stencilX': `${size.width / size.defaultWidth}`,
    '--stencilY': `${size.height / size.defaultHeight}`,
  };

  return (
    <Draggable bounds='parent' cancel='.Stencil-resize'>
      <div className='Stencil' style={style} ref={stencilRef}>
        {isEditingStencil ? <img
          className='Stencil-clear'
          src={close}
          alt=''
          aria-label='Clear the stencil'
          data-testid='stencil-close'
          onClick={handleClearDisplay}
        /> : null}
        {stencil.startsWith('data:image')
          ? <img className='Stencil-shape' src={stencil} alt='' />
          : <span className='Stencil-text'>{stencil}</span>
        }
        {isEditingStencil ? <img
          className='Stencil-resize'
          src={resizeIcon}
          draggable={false}
          alt=''
          aria-label='Resize the stencil'
          data-testid='stencil-resize'
          onMouseDown={beginResize}
          onMouseUp={endResize}
          onMouseMove={handleResize}
        /> : null}
      </div>
    </Draggable>
  );
};

export default Stencil;