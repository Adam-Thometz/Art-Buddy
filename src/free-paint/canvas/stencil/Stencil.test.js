import React from "react";

import rootReducer from "_redux/rootReducer";
import { setStencil } from "_redux/free-paint/freePaintActions";

import renderWithProvider from "_testUtils/renderWithProvider";

import Stencil from "./Stencil";

describe('Stencil component', () => {
  rootReducer.dispatch(setStencil('A'));

  it('renders without crashing', () => {
    renderWithProvider(<Stencil />, { store: rootReducer });
  });

  it('matches snapshot', () => {
    const { asFragment } = renderWithProvider(<Stencil />, { store: rootReducer });
    expect(asFragment()).toMatchSnapshot();
  });
});