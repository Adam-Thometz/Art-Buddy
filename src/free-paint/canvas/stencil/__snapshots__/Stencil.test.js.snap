// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stencil component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Stencil"
  >
    <span
      class="Stencil-text"
    >
      A
    </span>
  </div>
</DocumentFragment>
`;
